src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,757 @@
+            )
+            dy = np.divide(
+                np.multiply(dy_c1, c2) - np.multiply(dy_c2, c1), np.power(c2, 2)
+            )
+            dz = np.divide(
+                np.multiply(dz_c1, c2) - np.multiply(dz_c2, c1), np.power(c2, 2)
+            )
+
+            # update
+            x = x + np.multiply(alpha1, dx)
+            y = y + np.multiply(alpha2, dy)
+            z = z + np.multiply(alpha2, dz)
+
+            if iter_no < 2:
+                stop_cond = np.minimum(np.minimum(dx, dy), dz) / 1000
+
+            if iter_no > 400:
+                stop_cond = np.maximum(np.maximum(dx, dy), dz) + 1
+
+            iter_no = iter_no + 1
+
+        df1 = np.divide(x + np.abs(y + z * 1j), 2)
+        df2 = np.divide(x - np.abs(y + z * 1j), 2)
+        angle_ast = np.angle(y + z * 1j) / 2
+
+        inner_cosine = np.multiply(y, np.cos(2 * theta)) + np.multiply(
+            z, np.sin(2 * theta)
+        )
+        outer_sine = np.sin(np.multiply(a, x) + np.multiply(a, inner_cosine) - b)
+        outer_cosine = np.cos(np.multiply(a, x) + np.multiply(a, inner_cosine) - b)
+
+        sine_x_term = a
+        sine_y_term = np.multiply(a, np.cos(2 * theta))
+        sine_z_term = np.multiply(a, np.sin(2 * theta))
+
+        c1 = np.sum(np.multiply(np.abs(outer_sine), signal))
+        c2 = np.sqrt(np.multiply(sum_A, np.sum(np.power(outer_sine, 2))))
+
+        p = np.divide(c1, c2)
+
+        return df1, df2, angle_ast, p
+
+    def ctf_locate_minima(self, signal):
+        """
+        Find zero-crossings of power spectrum.
+
+        :param signal: Estimated power spectrum
+        :return: array of zero-crossings indices.
+        """
+
+        N = signal.shape[0]
+
+        derivatives = np.zeros((8, signal.shape[0], signal.shape[1]))
+        derivatives[0, 1:, :] = signal[: N - 1, :] - signal
+        derivatives[1, : N - 1, :] = signal[1:, :] - signal
+        derivatives[2, :, 1:] = signal[:, 1:] - signal
+        derivatives[3, :, : N - 1] = signal[:, 1:] - signal
+        derivatives[4, 1:, 1:] = signal[: N - 1, : N - 1] - signal
+        derivatives[5, 1:, : N - 1] = signal[: N - 1, 1:] - signal
+        derivatives[6, : N - 1, 1:] = signal[1:, : N - 1] - signal
+        derivatives[7, : N - 1, : N - 1] = signal[1:, 1:] - signal
+
+        derivative_sign = np.where(derivatives >= 0, 1, 0)
+        objective = np.sum(derivative_sign, axis=-1)
+        zero_cross_map = np.where(objective >= 6, 1, 0)
+
+        return zero_cross_map
+
+    # Note, This doesn't actually use anything from the class.
+    # It is used in a solver loop of some sort, so it may not be correct
+    # to just use what is avail in the obj.
+    def write_star(self, df1, df2, ang, cs, voltage, pixel_size, amp, name):
+        """
+        Writes starfile.
+        """
+
+        if os.path.isdir("results") is False:
+            os.mkdir("results")
+
+        f = open("results/" + os.path.splitext(name)[0] + ".log", "w")
```

Comment:
Should preferably use the tools we already have for writing STAR files (`aspire.storage.StarFile`).

Reply:
We can, but that code looks like a mess. It isn't clear at all how to use it (I'm looking at the UT now). Lets make that a separate issue from this PR.  Also I'm not sure who came first or does this better (ie https://github.com/alisterburt/starfile etc). Maybe we don't need to be in business of starfile wrappers of pandas...

Reply:
#412

Reply:
Fine with just leaving as is, then.